content/posts/2025-gsoc-clang-doc.md

```diff
@@ -0,0 +1,375 @@
+Future me was extremely thankful for the easy-to-use functions I had added.
+
+## Bitcode Refactor
+The bitcode read/write operations contain highly repetitive code.
+Adding something to documentation, like serializing `const` for a function, required several copy-pastes in several locations.
+It was structured like so:
+
+```cpp
+case BI_MEMBER_TYPE_BLOCK_ID: {
+  MemberTypeInfo TI;
+  if (auto Err = readBlock(ID, &TI))
+    return Err;
+  if (auto Err = addTypeInfo(I, std::move(TI)))
+    return Err;
+  return llvm::Error::success();
+}
+```
+
+`addTypeInfo` is specific for `MemberTypeInfo`, so every other type of `Info` would need to call its own function.
+Hence, highly repetitive similar code.
+I refactored that block to this:
+
+```cpp
+  return handleTypeSubBlock<TypeInfo>(ID, I, CreateAddFunc(addTypeInfo<T, TypeInfo>));
+```
+
+`handleTypeSubBlock` contains the same logic as the previous block, but it calls a generic `Function`.
+All of this was achieved without compromising the performance of documentation generation.
+
+## Mangling Filenames
+Clang-Doc had a bug stemming from non-unique filenames.
+The YAML backend avoided this problem because its filenames were SymbolIDs, but this meant that the lit tests would have to use regex to find the file for FileCheck.
+Nasty.
+
+In HTML and JSON, the filenames for classes were just the class name.
+If you had a template specialization, this would cause problems
+In HTML, we'd get duplicate HTML resulting in wonky web pages.
+In JSON, we'd get a fatal error from the JSON parser since there were two sets of top level braces.
+I used `ItaniumMangleContext` to generate mangled names we could use for the filenames.
+
+## Pull Requests
+
+Here are the pull requests I made for refactors during the project:
+
+- [Serialize record files with mangled name](https://github.com/llvm/llvm-project/pull/148021)
+- [refactor BitcodeReader::readSubBlock](https://github.com/llvm/llvm-project/pull/145835)
+- [refactor JSONGenerator array usage](https://github.com/llvm/llvm-project/pull/145595)
+- [refactor JSON for better Mustache compatibility](https://github.com/llvm/llvm-project/pull/149588)
+
+# Overview
+I implemented a new JSON generator that will serve as the basis for Clang-Doc's documentation generation.
+This will vastly reduce overall lines of code and maintenance burdens.
+I added a lot of tests to increase code coverage and ensure we are serializing all the information necessary for high-quality documentation.
+I refactored our comment handling to streamline the logic that handles them and for better output in the HTML.
+I also explored options for rendering Markdown and began an implementation for a parser that I plan on working on in the future.
+Along the way, I also did some refactoring to improve code reuse and improved maintenance burdens by reducing boilerplate code.
```

Comment:
I'd consider making this a list for better readability.
```suggestion
- I implemented a new JSON generator that will serve as the basis for Clang-Doc's documentation generation. This will vastly reduce overall lines of code and maintenance burdens.
- I added a lot of tests to increase code coverage and ensure we are serializing all the information necessary for high-quality documentation.
- I refactored our comment handling to streamline the logic that handles them and for better output in the HTML.
- I also explored options for rendering Markdown and began an implementation for a parser that I plan on working on in the future.
- Along the way, I also did some refactoring to improve code reuse and improved maintenance burdens by reducing boilerplate code.
```